requests.go

```diff
@@ -0,0 +1,294 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/http/cookiejar"
+	"net/url"
+)
+
+const (
+	scheme                = "https://"
+	loginPagePath         = "/portal/"
+	loginClearPath        = "/portal/selfservice/users/loginClear/"
+	loginSubmitPath       = "/portal/login"
+	loginOauth2Path       = "/login/sessionCookieRedirect?"
+	oauth2RedirectUrlPath = "/myportal/"
+	userDataPath          = "/api/v1/sessions/me/lifecycle/refresh"
+	changePasswordPath    = "/myportal/viewprofile/myprofile/credential"
+	logoutPath            = "/myportal/logout"
+	userId                = "userId"
+	sessionToken          = "sessionToken"
+)
+
+type loginDetails struct {
+	Username string          `json:"username"`
+	Password string          `json:"password"`
+	Options  map[string]bool `json:"options"`
+}
+
+type changePassword struct {
+	OldPassword string `json:"oldPassword"`
+	NewPassword string `json:"newPassword"`
+}
+
+type userData struct {
+	UserId       string `json:"userId"`
+	SessionToken string `json:"sessionToken"`
+}
+
+func getCookie(c *http.Client, urlstr, cookieName string) *http.Cookie {
+	urlObj, err := url.Parse(urlstr)
+	if err != nil {
+		panic(err)
+	}
+	cookies := c.Jar.Cookies(urlObj)
+	for _, c := range cookies {
+		if c.Name == cookieName {
+			return c
+		}
+	}
+	return nil
+}
+
+func sendRequest(client *http.Client, method, urlstr string, customHeaders map[string][]string, body []byte, userData interface{}) error {
+	var req *http.Request
+	var err error
+	if method == http.MethodGet {
+		req, err = http.NewRequest(http.MethodGet, urlstr, nil)
+	} else if method == http.MethodPost {
+		req, err = http.NewRequest(http.MethodPost, urlstr, bytes.NewReader(body))
+	} else {
+		return errors.New("unsupported method type")
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if len(body) > 0 {
+		req.Header.Add("Content-Type", contentType)
+	}
+
+	// Add common headers
+	req = addHeadersToRequest(req, commonHeaders)
+
+	// Add custom headers
+	req = addHeadersToRequest(req, customHeaders)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		b, err := io.ReadAll(resp.Body)
+		if err != nil {
+			log.Printf("failed to read response body: %s", err)
+		}
+		return fmt.Errorf("got HTTP status code %d with body: \n%s", resp.StatusCode, string(b))
+	}
+
+	if userData != nil {
+		err = json.NewDecoder(resp.Body).Decode(userData)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func loginStep(client *http.Client, portal *Portal, username, password string) error {
+	// GET login page at scheme+hostname+/portal/
+	hostname := portal.Hostname
+	idmHostname := portal.IDMHostname
+
+	headers := map[string][]string{
+		"Host":                      {hostname},
```

Comment:
I'm pretty sure Go's client library will ignore whatever you put as a Host header in the header map. If you want to override it you need to do it in the `Request` struct, but the default behavior (determine from URL) is what you want.

https://pkg.go.dev/net/http#Request

Reply:
ok. Removed `Host` header from map.